packages/react-next/src/components/FocusTrapZone/FocusTrapZone.tsx

```diff
@@ -8,334 +8,389 @@ import {
-    if (!disableFirstFocus && !elementContains(this._root.current, this._previouslyFocusedElementOutsideTrapZone)) {
-      this.focus();
+      : (getDocument(root.current)!.activeElement as HTMLElement);
+    if (!disableFirstFocus && !elementContains(root.current, state.previouslyFocusedElementOutsideTrapZone)) {
+      focus();
     }
-  }
-
-  private _returnFocusToInitiator(): void {
-    const { ignoreExternalFocusing } = this.props;
-
-    FocusTrapZone._focusStack = FocusTrapZone._focusStack.filter((value: FocusTrapZone) => {
-      return this !== value;
-    });
+  };
 
-    const doc = this._getDocument();
-    const activeElement = doc.activeElement as HTMLElement;
-    if (
-      !ignoreExternalFocusing &&
-      this._previouslyFocusedElementOutsideTrapZone &&
-      typeof this._previouslyFocusedElementOutsideTrapZone.focus === 'function' &&
-      (elementContains(this._root.current, activeElement) || activeElement === doc.body)
-    ) {
-      this._focusAsync(this._previouslyFocusedElementOutsideTrapZone);
+  const setAsyncFocus = (element: HTMLElement): void => {
+    if (!isBumper(element)) {
+      focusAsync(element);
     }
-  }
-
-  private _updateEventHandlers(newProps: IFocusTrapZoneProps): void {
-    const { isClickableOutsideFocusTrap = false, forceFocusInsideTrap = true } = newProps;
+  };
 
-    if (forceFocusInsideTrap && !this._disposeFocusHandler) {
-      this._disposeFocusHandler = on(window, 'focus', this._forceFocusInTrap, true);
-    } else if (!forceFocusInsideTrap && this._disposeFocusHandler) {
-      this._disposeFocusHandler();
-      this._disposeFocusHandler = undefined;
+  const returnFocusToInitiator = (): void => {
+    const { ignoreExternalFocusing } = props;
+    const doc = getDocument(root.current);
+
+    if (doc) {
+      const activeElement = doc.activeElement as HTMLElement;
+      if (
+        !ignoreExternalFocusing &&
+        state.previouslyFocusedElementOutsideTrapZone &&
+        typeof state.previouslyFocusedElementOutsideTrapZone.focus === 'function' &&
+        (elementContains(root.current, activeElement) || activeElement === doc.body)
+      ) {
+        setAsyncFocus(state.previouslyFocusedElementOutsideTrapZone);
+      }
     }
+  };
 
-    if (!isClickableOutsideFocusTrap && !this._disposeClickHandler) {
-      this._disposeClickHandler = on(window, 'click', this._forceClickInTrap, true);
-    } else if (isClickableOutsideFocusTrap && this._disposeClickHandler) {
-      this._disposeClickHandler();
-      this._disposeClickHandler = undefined;
-    }
-  }
+  const updateEventHandlers = (): void => {
+    const { isClickableOutsideFocusTrap = false, forceFocusInsideTrap = true } = props;
+    const win = getWindow(root.current)!;
 
-  private _onFocusCapture = (ev: React.FocusEvent<HTMLDivElement>) => {
-    if (this.props.onFocusCapture) {
-      this.props.onFocusCapture(ev);
+    if (forceFocusInsideTrap && !state.disposeFocusHandler) {
```

Comment:
Is there any way we can handle click and focus handler creation and disposal with effects or something? While a 1-1 mapping from private members to internal state may technically work, it would be nice if we can take advantage of automatic updating and disposal patterns provided by hooks.